docs/ocis/adr/0021-service-authorization.md

```diff
@@ -0,0 +1,80 @@
+---
+title: "21. Service authorization"
+date: 2023-01-18T16:07:00+01:00
+weight: 21
+geekdocRepo: https://github.com/owncloud/ocis
+geekdocEditPath: edit/master/docs/ocis/adr
+geekdocFilePath: 0020-service-authorization.md
+---
+
+* Status: proposed
+* Deciders: @butonic
+* Date: 2023-01-18
+
+## Context and Problem Statement
+
+Some services need access to file content without a user being logged in. We currently pass the owner or manager
+of a space in events which allows the search service to impersonate that user to extract metadata from the changed resource.
+There are two problems with this:
+1. The service could get all permissions of the user and gain write permission
+2. There is a race condition where the user in the event might no longer have read permission, causing the index to go stale
+
+The race condition will become more of an issue when we start working on a workflow engine.
+
+How can we grant services the least amount of permissions required for their purpose?
+
+## Decision Drivers
+
+* It should be possible to represent this as servicePrincipals in the libregraph API, similar to the [MS Graph servicePrincipal](https://learn.microsoft.com/en-us/graph/api/resources/serviceprincipal?view=graph-rest-1.0).
+* Services should check permissions using the ocis permissions or reva auth service, we don't want to introduce a new mechanism for this
+
+## Considered Options
+
+* [Service Accounts](#service-accounts)
+* [Auth Manager type Space-Owner](#auth-manager-type-space-owner)
+
+## Decision Outcome
+
+Chosen option: ?
+
+### Positive Consequences
+
+* ???
+* ???
+* ???
+
+### Negative consequences
+
+* ???
+* ???
+* ???
+
+## Pros and Cons of the Options
+
+### Service Accounts
+
+Implement service accounts to mimic the MS graph servicePrincipals for an application. The appRoles can be used to limit the interactions to other services, e.g. `Files.Read.All` to allow the servicePrincipal reading all files in all drives of a tenant. 
+An application is represented by a dedicated servicePrincipal per tenant, which is created when granting an application access to a tenant (currently instace) by an admin. These servicepPrancipals then get an appRoleAssignment for the appRoles defined by the application.
+An application could be the search service (or only a content extraction service), a workflow service, a publication service or any other service.
+
+TODO which services need a service account?
```

Comment:
We can already start composing a list:
```
search, audit, notifications, thumbnails
```

Reply:
Why does `thumbnails` need a service account? It does not listen to events but creates thumbnails on the fly. It can do that with the logged in user. In the graph api this maps to delegated permissions, where the user grants a service one of his permissions to act on his behalf. Or the admin grants a delegated permission on behalf of all users. But that would mean that we would have to build a request context that contains not only the logged in user, but also the servicePrancipal. Only then could the permission system evaluate a `Files.Read.All` delegated permission granted to the servicePrincipal by taking the logged in user into account.

To clarify, ms graph also distinguishes between [application and delegated permissions](https://learn.microsoft.com/en-us/graph/permissions-overview?tabs=http#comparison-of-delegated-and-application-permissions):

Permission name | Type | ID
-- | -- | --
Files.ReadWrite.All | Application | 75359482-378d-4052-8f01-80520e7db3cd
`Files.ReadWrite.All` | Delegated | 863451e7-0667-486c-a5d6-d135439485f0

In any case, this would allow restricting the permissions a service has even further. We could ggive the thumbnails service the `Files.ReadWrite.All` delegated permission. AFAICT we would have to extend the CS3 `CheckPermissionRequest` to contain a servicePrincipal property.